src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+        {
+            var generator = editor.Generator;
+            var arguments = generator.CreateArguments(methodSymbol.Parameters);
+
+            // Remove trailing trivia because it adds spaces to the beginning of the following statement.
+            arguments = arguments.Add(generator.Argument(expression.WithoutTrailingTrivia()));
+            var memberName = methodSymbol.IsGenericMethod
+                ? generator.GenericName(methodSymbol.Name, methodSymbol.TypeArguments)
+                : generator.IdentifierName(methodSymbol.Name);
+
+            var invocation = generator.InvocationExpression(memberName, arguments);
+            var newStatements = SpecializedCollections.SingletonEnumerable(invocation);
+
+            if (!methodSymbol.ReturnsVoid)
+            {
+                newStatements = SpecializedCollections.SingletonEnumerable(generator.ReturnStatement(invocation));
+            }
+
+            var newMethod = CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol);
+            var newMethodDeclaration = generator.MethodDeclaration(newMethod, newStatements);
+            editor.InsertBefore(containingMethod, newMethodDeclaration);
+        }
+
+        /// <summary>
+        /// Gets the matches of the expression and replaces them with the identifier.
+        /// Special case for the original matching expression, if its parent is a LocalDeclarationStatement then it can
+        /// be removed.
+        /// Otherwise, it needs to have a rename annotation added to it.
+        /// </summary>
+        public static async Task UpdateExpressionInOriginalFunctionAsync(Document document,
+            TExpressionSyntax expression, SyntaxNode scope, string parameterName, SyntaxEditor editor,
+            bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var matches = await FindMatchesAsync(document, expression, scope, allOccurrences, cancellationToken).ConfigureAwait(false);
+            var replacement = (TIdentifierNameSyntax)generator.IdentifierName(parameterName);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            foreach (var match in matches)
+            {
+                // Special case the removal of the originating expression to either remove the local declaration
+                // or to add a rename annotation.
+                if (!match.Equals(expression))
+                {
+                    editor.ReplaceNode(match, replacement);
+                }
+                else
+                {
+                    if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName))
+                    {
+                        var localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
```

Comment:
this may not work and can crash.  as an example, if you ahve `for (var x = ...;`  then that's a vardecl that doesn't have a containing local decl statement.  please add test and revise.

This would be much simpler if ShouldRemoveVariableDeclaratorContainingExpression actually verifies (and returns) the local decl it wants to remove.

Reply:
I will change it to return the local decl. I assume I can just get the name of the variable by descending and finding the variable declarator?

Reply:
not necessarily.  and there are cases like this:   `int x = a * b, y = 0;`  def check for this.  only return the name *and* declaration in the single case.

Reply:
got it